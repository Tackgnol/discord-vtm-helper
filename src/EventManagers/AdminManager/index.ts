import { globalTestRichEmbedInit, narrationRichEmbed, npcRichEmbed, playerRichEmbed, statInsightRichEmbed } from '../../Common';
import { settings } from '../../config/settings';
import {
	isObject,
	validateAddFacts,
	validateAddNarration,
	validateAddNPC,
	validateAddPlayer,
	validateAddStatInsight,
} from './validators';
import { trim } from 'lodash';
import { IStat } from '../../Models/GameData';
import { errorName, InvalidInputError } from '../../Common/Errors';
import { IReply, ReplyType } from '../../Models/AppModels';
import { MessageEmbed } from 'discord.js';

export class AdminManager {
	async fireEvent(eventName: string, value: string, gameId: string, channelId: string, authorId: string): Promise<IReply> {
		const {
			addPlayer,
			addNPC,
			addFact,
			removePlayer,
			addGlobalTest,
			addStatInsight,
			addNarration,
			assignAdmin,
		} = settings.subPrefixes.adminSubCommands;
		let result: MessageEmbed | string | null;
		switch (eventName) {
			case addPlayer:
				result = await this.addPlayer(value, gameId);
				break;
			case addNPC:
				result = await this.addNPC(value, gameId);
				break;
			case addFact:
				result = await this.addFactsToNPCs(value, gameId);
				break;
			case removePlayer:
				throw Error('Unimplemented');
			case addGlobalTest:
				result = await this.addGlobalTest(value, channelId, gameId);
				break;
			case addStatInsight:
				result = await this.addStatInsight(value, channelId, gameId);
				break;
			case addNarration:
				result = await this.addNarrationEvent(value, channelId, gameId);
				break;
			case assignAdmin:
				result = await this.assignAdminToChannel(authorId, channelId);
				break;
			default:
				throw new InvalidInputError('Invalid admin command!');
		}
		if (!result) {
			throw new InvalidInputError('No reply available');
		}
		return { type: ReplyType.Personal, value: result };
	}

	private async addPlayer(value = '', channelId: string) {
		const addPlayerRegex = /\[([^\]]+)\].*\[([^\]]+)\].*\[([^\]]+)\]/g;
		const parsed = addPlayerRegex.exec(value);
		let statArray: IStat[] = [];
		if (!parsed || parsed.length < 4) {
			return 'Invalid input, was expecting [player discord name][player discord id][statistics input]';
		} else {
			const errorArray = validateAddPlayer(parsed);
			const name = parsed[1];
			const id = parsed[2];
			const stats = parsed[3];
			if (errorArray.length >= 1) {
				return errorArray.toString();
			} else if (isObject(stats)) {
				const onlyStatsRegex = /{([\w+:\d, ?]+[\w+:\d]{1})}/g;
				const matchStats = onlyStatsRegex.exec(stats);
				const onlyStats = matchStats ? matchStats[1].split(',') : [];
				statArray = onlyStats.map(stat => {
					const split = stat.split(':');
					return { name: split[0], value: +split[1] };
				});
			} else {
				const { statList } = settings;
				const statValues = stats.split(',');
				for (let i = 0; i < statValues.length; i++) {
					const statName = statList[i] ? statList[i] : '';
					const statValue = statValues[i] ? statValues[i] : 0;
					statArray.push({ name: statName, value: +statValue });
				}
			}
			const result = await global.service.AddPlayer(name, id, statArray, channelId);
			if (result) {
				return playerRichEmbed(name, id, statArray);
			} else {
				return 'Something went wrong while saving your character...';
			}
		}
	}

	private async addNPC(value = '', gameId: string) {
		const addNPCRegex = /\[(\w+)\].*\[(\w+)\].*\[(.+)\]\[(\w.+)\]/g;
		const parsed = addNPCRegex.exec(value);
		if (!parsed || parsed.length < 5) {
			return 'Invalid input, was expecting [npc name][npc call name][npc image][npc description]';
		} else {
			const errorArray = validateAddNPC(parsed);
			if (errorArray.length > 0) {
				return errorArray.toString();
			}
			const name = parsed[1];
			const callName = parsed[2];
			const image = parsed[3];
			const description = parsed[4];
			const result = await global.service.AddNPC(name, callName, image, description, gameId);
			if (result) {
				return npcRichEmbed(result, true, true);
			} else {
				return 'Something went wrong while saving your npc...';
			}
		}
	}

	private async addFactsToNPCs(value = '', gameId: string) {
		const addFactRegex = /\[(\w+)\]\[([\d+,]+\d+)\]\[(.+)\]/g;
		const parsed = addFactRegex.exec(value);
		const results = [];
		if (!parsed || parsed.length < 4) {
			return 'Invalid input was expecting [npc call name][player discord ids][fact list]';
		} else {
			const errorArray = validateAddFacts(parsed);
			if (errorArray.length > 0) {
				return errorArray.toString();
			}
			const callName = parsed[1];
			const playerList = parsed[2] ? parsed[2].split(',').map(p => trim(p)) : [];
			const factList = parsed[3] ? parsed[3].split(',').map(f => trim(f)) : [];
			for (const p of playerList) {
				const result = await global.service.AddFactsToNPC(p, callName, factList, gameId);
				results.push(result);
			}
			return results.toString();
		}
	}

	private async addNarrationEvent(value = '', channelId: string, gameId: string) {
		const addNarrationRegex = /\[(\w+)\]\[(.+)\]\[(.+)\]/g;
		const parsed = addNarrationRegex.exec(value);
		if (!parsed || parsed.length < 4) {
			return 'Invalid input was expecting [event call name][image to display][Event description]';
		} else {
			const errorArray = validateAddNarration(parsed);
			if (errorArray.length > 0) {
				return errorArray.toString();
			}
			const callName = parsed[1];
			const image = parsed[2];
			const description = parsed[3];
			const result = await global.service.AddNarration(callName, image, description, channelId, gameId);
			if (result) {
				return narrationRichEmbed(result.narrationText, result.image, true);
			} else {
				return null;
			}
		}
	}

	private async addStatInsight(value = '', channelId: string, gameId: string) {
		const addStatInsightRegex = /\[([A-Za-z]+)\]\[([A-Za-z]+)\]\[(\d)\]\[([A-Za-z ,;'"\\s]+[.?!]?)\]/g;
		const parsed = addStatInsightRegex.exec(value);
		if (!parsed || parsed.length < 5) {
			return 'Invalid input was expecting [event call name][Stat to check][stat value][Success message]';
		} else {
			const errorArray = validateAddStatInsight(parsed);
			if (errorArray.length > 0) {
				return errorArray.toString();
			}
			const eventName = parsed[1];
			const statName = parsed[2];
			const statValue = +parsed[3];
			const successMessage = parsed[4];
			const result = await global.service.AddStatInsight(eventName, statName, statValue, successMessage, channelId, gameId);
			if (result) {
				return statInsightRichEmbed(statName, +statValue, successMessage, true);
			} else {
				return null;
			}
		}
	}

	private async addGlobalTest(value = '', channelId: string, gameId: string) {
		const addGlobalTestRegex = /\[([\w\d ]+)\]\[([aA-zZ0-9 .,?!]+)\]\[(true|false)\]\[([aA-zZ0-9 .,?!]+)\]\[(({\w+: ?\d ?, ?\w+: ?"[aA-zZ0-9 !?,.]+" ?}, )+({\w+:\d ?, ?\w+:? "[aA-zZ0-9 !?,.]+" ?}))\]/g;
		const parsed = addGlobalTestRegex.exec(value);
		if (!parsed || parsed.length < 6) {
			return 'Invalid input was expecting [event call name][Test message][true or false][Message to display before the result][options for results: {minResult:value, resultMessage: "message for the player that achives this result" }]';
		} else {
			const eventName = parsed[1];
			const testMessage = parsed[2];
			const shortCircuit = parsed[3];
			const replyPrefix = parsed[4];
			const options = parsed[5];
			const optionRegex = /(minResult: ?\d, ?resultMessage: ?"[aA-zZ !?,.]+" ?)/g;
			const messageRegex = /"(.+)"/g;
			const optionsParsed = options.match(optionRegex);
			const optionsArray = optionsParsed?.map(o => {
				const split = o.split(',');
				const minResult = split[0].replace('minResult:', '').trim();
				const messageRegexResult = split[1].match(messageRegex);
				const message = messageRegexResult ? messageRegexResult[0] : '';
				return {
					minResult: +minResult,
					resultMessage: message,
				};
			});

			return global.service
				.AddGlobalTest(eventName, testMessage, Boolean(shortCircuit), replyPrefix, optionsArray ?? [], channelId, gameId)
				.then(() => {
					return globalTestRichEmbedInit(testMessage, true);
				})
				.catch((e: Error) => {
					if (e.name === errorName) {
						return 'This name already exists, specify a new one';
					} else {
						return 'Failed to add event';
					}
				});
		}
	}

	private async assignAdminToChannel(authorId: string, channelId: string) {
		return global.service
			.AssignGameAdmin(authorId, channelId, '216a9908-2766-42cc-be08-ea717593447a')
			.then(() => {
				return `Congrats you are now an admin of ${channelId}!`;
			})
			.catch(e => {
				return `Adding admin to channel failed\n${e}`;
			});
	}
}
