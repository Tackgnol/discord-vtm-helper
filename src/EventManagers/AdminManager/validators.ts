const imageRegex = /(https?:\/\/.*\.(?:png|jpg|jpeg))/i;

const validateAddPlayer = (input: RegExpExecArray) => {
	const errorArray = [];
	if (!(typeof input[1] === 'string')) {
		errorArray.push('DiscordClient name not specified');
	}
	if (!(typeof input[2] === 'string')) {
		errorArray.push('DiscordId not specified');
	}
	if (!input[3]) {
		errorArray.push('Statistics array / object not filled');
	} else if (input[3].charAt(0) === '{' && input[3].slice(-1) !== '}') {
		errorArray.push('Invalid stat object format, please open and close the {}');
	} else if (input[3].charAt(0) === '{' && input[3].slice(-1) === '}' && !/{(\w+:\d, ?)+(\w+:\d){1}}/g.test(input[3])) {
		errorArray.push('Invalid stat format expected {stat:value,stat:value,stat:value}');
	} else if (input[3].charAt(0) !== '{' && !/(\d,)+(\d){1}/.test(input[3])) {
		errorArray.push('Invalid stat format expected value,value,value');
	}
	return errorArray;
};

const validateAddNPC = (input: RegExpExecArray) => {
	const errorArray = [];
	if (!(typeof input[1] === 'string')) {
		errorArray.push('NPC name not specified');
	}
	if (!(typeof input[2] === 'string')) {
		errorArray.push('NPC call name not specified');
	}
	if (!(typeof input[3] === 'string')) {
		errorArray.push('NPC image not specified');
	} else if (!imageRegex.test(input[3])) {
		errorArray.push('NPC image url invalid, accepted formats: jpg, jpeg and png');
	}
	if (!input[4]) {
		errorArray.push('NPC description not specified');
	}
	return errorArray;
};

const validateAddFacts = (input: RegExpExecArray) => {
	const errorArray = [];
	const IdSeparateChecker = /^([0-9]+,?\s*)+$/g;
	const factSeparateChecker = /^([a-zA-Z0-9]+,?\s*)+$/g;
	if (!(typeof input[1] === 'string')) {
		errorArray.push('npc name not specified');
	}
	if (!IdSeparateChecker.exec(input[2])) {
		errorArray.push('Invalid playerIds insert playerId seperated by commas ex. [1241512, 1512521,123123]');
	}
	if (!factSeparateChecker.exec(input[3])) {
		errorArray.push('Invalid facts insert facts separated by commas ex. [Fact 1, Fact 2,Fact3]');
	}
	return errorArray;
};

const validateAddNarration = (input: RegExpExecArray) => {
	const errorArray = [];
	if (!(typeof input[1] === 'string')) {
		errorArray.push('Event name must a text');
	} else if (/^\w+%/i.exec(input[1])) {
		errorArray.push('Only letters are allowed in the event name, no spaces, no special characters');
	}
	if (!(typeof input[2] === 'string')) {
		errorArray.push('Event image must a text');
	} else if (!imageRegex.test(input[2])) {
		errorArray.push('NPC image url invalid, accepted formats: jpg, jpeg and png');
	}
	if (!(typeof input[3] === 'string')) {
		errorArray.push('Event description must a text');
	}
	return errorArray;
};

const validateAddStatInsight = (input: RegExpExecArray) => {
	const errorArray = [];
	if (!(typeof input[1] === 'string')) {
		errorArray.push('Event name must a text');
	} else if (/^\w+%/i.exec(input[1])) {
		errorArray.push('Only letters are allowed in the event name, no spaces, no special characters');
	}
	if (!(typeof input[2] === 'string')) {
		errorArray.push('stat name must a text');
	} else if (/^\w+%/i.exec(input[2])) {
		errorArray.push(
			'Only letters are allowed in the stat name, no spaces, no special characters, if a stat has a space please use CamelCase'
		);
	}
	if (isNaN(+input[3])) {
		errorArray.push('Stat value needs to be a number');
	}
	if (!(typeof input[4] === 'string')) {
		errorArray.push('Success message needs to be string');
	}
	return errorArray;
};

const isObject = (input: string) => {
	return input.charAt(0) === '{' && input.slice(-1) === '}';
};

export { isObject, validateAddFacts, validateAddNPC, validateAddPlayer, validateAddNarration, validateAddStatInsight };
