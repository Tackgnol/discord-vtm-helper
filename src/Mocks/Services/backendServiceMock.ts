import { IService } from '../../Services/IService';
import { GlobalTest, Narration, NPC, Player, Stat, StatInsight, Option } from '../../Models/GameData';
import { Game, SessionData } from '../../Models/AppModels';
import { testNPCWithFacts, testNPCWithoutFacts } from '../SessionDataMocks/NPCMocks';
import { globalTestWithShortCircuitMock } from '../SessionDataMocks/GlobalTestMocks';
import { basicNarration } from '../SessionDataMocks/NarrationMocks';
import { playerMock } from '../SessionDataMocks/PlayerMock';
import { statInsightMockLow } from '../SessionDataMocks/StatInsightMocks';
import { activeSessionMock } from '../SessionDataMocks/ActiveSessionMock';
import { gameMock } from '../GameMock';

export class BackendServiceMock implements IService {
	AddFactsToNPC(playerId: string, npc: string, facts: string[], gameId: string): Promise<NPC> {
		if (!(playerId || npc || facts || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve({ ...testNPCWithFacts, facts: [...testNPCWithFacts.facts, ...facts] });
	}

	AddGlobalTest(
		name: string,
		message: string,
		shortCircuit: boolean,
		replyPrefix: string,
		globaltestoptionSet: Option[],
		channelId: string,
		gameId: string
	): Promise<GlobalTest> {
		if (!(name || message || shortCircuit || replyPrefix || globaltestoptionSet || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(globalTestWithShortCircuitMock);
	}

	AddNPC(name: string, callName: string, image: string, description: string, gameId: string): Promise<Omit<NPC, 'facts'>> {
		if (!(name || callName || image || description || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(testNPCWithoutFacts);
	}

	AddNarration(name: string, image: string, narrationText: string, channelId: string, gameId: string): Promise<Narration> {
		if (!(name || image || narrationText || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(basicNarration);
	}

	AddPlayer(name: string, id: string, statArray: Stat[], gameId: string): Promise<Player> {
		if (!(name || id || statArray || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	AddStatInsight(
		name: string,
		stat: string,
		value: number,
		message: string,
		channelId: string,
		gameId: string
	): Promise<StatInsight> {
		if (!(name || stat || value || message || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(statInsightMockLow);
	}

	AssignGameAdmin(playerId: string, channelId: string, gameId: string): Promise<SessionData> {
		if (!(playerId || channelId || gameId)) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetEvents(channelId: string, gameId: string): Promise<SessionData> {
		if (!(channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetPlayer(playerId: string, gameId: string): Promise<Player> {
		if (!(playerId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	GetUserChannels(userId: string): Promise<Game[]> {
		if (!userId) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve([gameMock]);
	}
}
