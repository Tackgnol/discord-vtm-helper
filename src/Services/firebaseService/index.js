const firebase = require('firebase/app');
const auth = require('../../../config/auth')
class FirebaseService {
	constructor() {
		this.service = firebase.initializeApp(auth.firebase);
	}
}

module.exports = FirebaseService;
