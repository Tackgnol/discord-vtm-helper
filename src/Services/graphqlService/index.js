const { ApolloClient } = require('apollo-boost');

const { createHttpLink } = require('apollo-link-http');
const { InMemoryCache } = require('apollo-cache-inmemory');
const fetch = require('node-fetch');
const settings = require('../../../config/settings');
const GET_CHANNELS = require('../../GraphQL/GET_CHANNELS');
const GET_NPCS = require('../../GraphQL/GET_NPCS');
const { find, get } = require('lodash');

class GraphqlService {
	constructor() {
		this.apolloClient = new ApolloClient({
			link: createHttpLink({ uri: settings.onlineSourceUrl, fetch: fetch }),
			cache: new InMemoryCache(),
		});
	}

	async GetPlayer(playerId) {
		const graphQLQuery = await this.apolloClient.query({
			query: GET_NPCS,
			variables: { playerId: playerId },
		});
		return get(graphQLQuery, 'data.player');
	}

	async GetEvents(channelId) {
		const graphQLQuery = await this.apolloClient.query({
			query: GET_CHANNELS,
		});
		return find(get(graphQLQuery, 'data.allChannels'), ed => ed.discordId === channelId);
	}
}

module.exports = GraphqlService;
