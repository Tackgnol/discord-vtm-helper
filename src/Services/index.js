const LocalFileService = require('./localFileService');
const GraphqlService = require('./graphqlService');

const initializeService = (serviceType) => {
	switch (serviceType) {
	case 'offline':
		return new LocalFileService();
	case 'graphql':
		return new GraphqlService();
	default:
		throw new Error('Invalid service type in config!');
	}
};

module.exports = initializeService;
