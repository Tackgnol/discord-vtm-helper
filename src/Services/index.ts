import LocalFileService from './localFileService';
import GraphqlService from './graphqlService';
import FirebaseService from './firebaseService';
import { IService } from './IService';

const initializeService = (serviceType: string): IService => {
	switch (serviceType) {
		case 'offline':
			return new LocalFileService();
		case 'graphql':
			return new GraphqlService();
		case 'firestore':
			return new FirebaseService();
		default:
			throw new Error('Invalid service type in config!');
	}
};

export default initializeService;
