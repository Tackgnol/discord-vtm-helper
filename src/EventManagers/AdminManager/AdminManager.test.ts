import { AdminManager } from './AdminManager';
import { BackendServiceMock } from '../../Mocks/Services/backendServiceMock';
import {
	faileAddPlayerNameOnlyQuery,
	failedAddFact,
	failedAddFactMissingFact,
	failedAddFactNameOnly,
	failedAddGlobalTest,
	failedAddGlobalTestMissingOptions,
	failedAddGlobalTestMissingOptionsAndPrefix,
	failedAddGlobalTestNameOnly,
	failedAddNarrationMissingMessage,
	failedAddNarrationNameOnly,
	failedAddNPC,
	failedAddNPCMissingDescription,
	failedAddNPCMissingDescriptionAndImage,
	failedAddNPCNameOnly,
	failedAddPlayerMissingQuery,
	failedAddPlayerQuery,
	failedAddStatInsightMissingMessage,
	failedAddStatInsightMissingStat,
	failedAddStatInsightNameOnly,
	successfulAddFact,
	successfulAddPlayerQueryList,
	successfulAddPlayerQueryObject,
	successfulAssignAdmin,
	successfullAddNPC,
	successfullyAddGlobalTest,
	successfullyAddNarration,
	successfullyAddStatInsight,
} from '../../Mocks/QueryMocks/adminQueries';
import chai, { expect } from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';

chai.use(jestSnapshotPlugin());

describe('EventManagers >>  AdminManager >> AdminManager', () => {
	const service = new BackendServiceMock();
	const manager = new AdminManager(service);

	it('successfully adds a player with a serial stat', async () => {
		const { value, eventName } = successfulAddPlayerQueryList;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		expect(result.value).to.be.not.null;
		expect(result).toMatchSnapshot();
	});

	it('successfully adds a player with a object stat', async () => {
		const { value, eventName } = successfulAddPlayerQueryObject;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		expect(result).toMatchSnapshot();
	});

	it('throws an error while missing stats', () => {
		const { value, eventName } = failedAddPlayerMissingQuery;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('throws an error on invalid value', () => {
		const { value, eventName } = failedAddPlayerQuery;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('throws an error only name is given', () => {
		const { value, eventName } = faileAddPlayerNameOnlyQuery;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('Successfully adds a npc with a proper query', () => {
		const { value, eventName } = successfullAddNPC;
		const result = manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result).to.not.be.null;
	});

	it('Fails to add a npc with a wrong query', () => {
		const { value, eventName } = failedAddNPC;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('Fails to add a npc with a missing description', () => {
		const { value, eventName } = failedAddNPCMissingDescription;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('Fails to add a npc with a missing description and image', () => {
		const { value, eventName } = failedAddNPCMissingDescriptionAndImage;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('Fails to add a npc with a wrong query', () => {
		const { value, eventName } = failedAddNPCNameOnly;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', 'x', '0', '0');
		}).to.throw;
	});

	it('should add a Fact on a proper input', async () => {
		const { value, eventName } = successfulAddFact;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result.value).length.above(0);
	});

	it('should fail to add a Fact on a invalid input', async () => {
		const { value, eventName } = failedAddFact;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('should fail to add a Fact on a missing fact list', async () => {
		const { value, eventName } = failedAddFactMissingFact;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('should fail to add a Fact on name only', async () => {
		const { value, eventName } = failedAddFactNameOnly;
		expect(() => {
			manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Successfully adds a new global test on a proper input', async () => {
		const { value, eventName } = successfullyAddGlobalTest;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result.value).length.above(0);
	});

	it('Fails to add a new global test on a invalid input', async () => {
		const { value, eventName } = failedAddGlobalTest;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new global test on a invalid input - missing options', async () => {
		const { value, eventName } = failedAddGlobalTestMissingOptions;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new global test on a invalid input - missing options and prefix', async () => {
		const { value, eventName } = failedAddGlobalTestMissingOptionsAndPrefix;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new global test on a invalid input - name only', async () => {
		const { value, eventName } = failedAddGlobalTestNameOnly;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Successfully adds a stat insight on a proper input', async () => {
		const { value, eventName } = successfullyAddStatInsight;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result.value).length.above(0);
	});

	it('Fails to add a new stat insight on a invalid input - missing message', async () => {
		const { value, eventName } = failedAddStatInsightMissingMessage;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new stat insight on a invalid input - missing message and stats', async () => {
		const { value, eventName } = failedAddStatInsightMissingStat;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new stat insight on a invalid input - name only', async () => {
		const { value, eventName } = failedAddStatInsightNameOnly;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Successfully adds  a Narration on proper input', async () => {
		const { value, eventName } = successfullyAddNarration;
		const result = await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result.value).length.above(0);
	});

	it('Fails to add a new Narration on a invalid input - missing message', async () => {
		const { value, eventName } = failedAddNarrationMissingMessage;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('Fails to add a new Narration on a invalid input - name only', async () => {
		const { value, eventName } = failedAddNarrationNameOnly;
		expect(async () => {
			await manager.fireEvent(eventName ?? '', value ?? '', '0', '0', '0');
		}).to.throw;
	});

	it('assigns the admin to a channel', async () => {
		const { eventName } = successfulAssignAdmin;
		const result = await manager.fireEvent(eventName ?? '', '', '0', '0', '0');
		expect(result).toMatchSnapshot();
		expect(result.value).length.above(0);
	});
});
