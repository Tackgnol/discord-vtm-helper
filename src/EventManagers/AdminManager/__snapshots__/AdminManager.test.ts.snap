// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventManagers >>  AdminManager >> AdminManager Successfully adds  a Narration on proper input 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "this is a narration text",
    "fields": Array [],
    "footer": null,
    "image": Object {
      "url": "urlToImage",
    },
    "thumbnail": null,
    "timestamp": null,
    "title": "Successfully added event: Silence, as I show what thy must see!",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager Successfully adds a new global test on a proper input 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "this is a new test",
    "fields": Array [],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Successfully added event",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager Successfully adds a npc with a proper query 1`] = `Promise {}`;

exports[`EventManagers >>  AdminManager >> AdminManager Successfully adds a stat insight on a proper input 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "This feels funny but you dont know what causes this yet",
    "fields": Array [
      Object {
        "inline": false,
        "name": "Occult",
        "value": "4",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Successfully added event",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager assigns the admin to a channel 1`] = `
Object {
  "type": 0,
  "value": "Congrats you are now an admin of 0!",
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager should add a Fact on a proper input 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 5832961,
    "description": null,
    "fields": Array [
      Object {
        "inline": false,
        "name": "Player: 0",
        "value": "fact1,fact2,test,test2",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Successfully added facts to a npc",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager successfully adds a player with a object stat 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "DiscordId: 409408071601881089",
    "fields": Array [
      Object {
        "inline": false,
        "name": "Occult",
        "value": "1",
      },
      Object {
        "inline": false,
        "name": "Investigation",
        "value": "2",
      },
      Object {
        "inline": false,
        "name": "Brawl",
        "value": "3",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Dudette",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >>  AdminManager >> AdminManager successfully adds a player with a serial stat 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "DiscordId: 409408071601881089",
    "fields": Array [
      Object {
        "inline": false,
        "name": "Strength",
        "value": "1",
      },
      Object {
        "inline": false,
        "name": "Dexterity",
        "value": "2",
      },
      Object {
        "inline": false,
        "name": "Stamina",
        "value": "3",
      },
      Object {
        "inline": false,
        "name": "Charisma",
        "value": "4",
      },
      Object {
        "inline": false,
        "name": "Manipulation",
        "value": "5",
      },
      Object {
        "inline": false,
        "name": "Composure",
        "value": "6",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Dudette",
    "type": "rich",
    "url": null,
  },
}
`;
